modal/runner/api.py

```diff
@@ -0,0 +1,80 @@
+import modal
+from fastapi import Depends, FastAPI, Request
+from fastapi.responses import JSONResponse
+from fastapi.security import HTTPAuthorizationCredentials
+from pydantic import BaseModel
+
+from runner.endpoints.completion import completion as completion_endpoint
+from runner.shared.common import config
+from runner.shared.download import download_model
+from shared.logging import get_logger
+from shared.protocol import CompletionPayload
+
+api_app = FastAPI()
+logger = get_logger(__name__)
+
+
+@api_app.middleware("http")
+async def log_errors(request: Request, call_next):
```

Comment:
when is this called?

Reply:
The fastapi middleware decorator here registers the fn to be called after every req's execution

Not sure off the top of my head how it behaves with a streaming endpoint, but I assume it's on or before a connection close / last bytes sent